### DISCUSSION TITLE
Clipboard.GetDataAsync returns null

### DISCUSSION THREAD

**Discussion by DirkKramer** [AvaloniaUI/Avalonia#12664] [Q&A]
My Clipboard.GetDataAsync  always returns null when i try to get an object from it.

this is my code

```
[RelayCommand]
async Task PasteConfigFromClipboard()
{
    if (OperatingSystem.IsAndroid())
        return;
    var _config = await (Application.Current as App).GetClipboard().GetDataAsync("model_to_copy");
    Configs.Add((VisionConfigurationEntity)_config);
}

[RelayCommand]
async Task CopyConfigToClipboard()
{
    if (OperatingSystem.IsAndroid())
        return;
    var model_to_copy = GetDuplicate();
    var dataObject = new DataObject();
    dataObject.Set("model_to_copy", model_to_copy);
    await (Application.Current as App).GetClipboard().SetDataObjectAsync(dataObject);
}
```

Environment:
Windows 11
Avalonia 11.0.1

**Comment by timunie** [comment]
Isn't Clipboard now part of TopLevel? 
https://docs.avaloniaui.net/docs/next/concepts/toplevel#clipboard

If you need to access if from the VM, use either of these approaches (both drafts): 
https://github.com/timunie/Avalonia.Samples/tree/feature/MVVM_Dialogs/src/Avalonia.Samples/MVVM/AdvancedMvvmDialogSample
https://github.com/timunie/Avalonia.Samples/tree/feature/MVVM_Dialogs/src/Avalonia.Samples/MVVM/MvvmDialogSample

**Reply by DirkKramer** [reply]
My clipboard works fine, although your approach is better.   this is in my App.axaml.cs to get the clipboard

```
public IClipboard GetClipboard()
{
    if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
    {      
         return TopLevel.GetTopLevel(desktop.MainWindow)?.Clipboard;
    }
    else if (ApplicationLifetime is ISingleViewApplicationLifetime singleViewPlatform)
    {       
        return TopLevel.GetTopLevel(singleViewPlatform.MainView)?.Clipboard;
    }
    return null;
}
```

The problem is that it gets null back when i ask it for the copied dataobject

**Comment by timunie** [comment]
dataobject is a byte array and thus needs to be handled accordingly. A direct cast via "as myObject" will result in `null` or throw.

**Comment by kekekeks** [comment]
We currently don't provide "the returned data object is the same that was passed to clipboard" guarantee. That's a part of [the  planned clipboard rework](https://github.com/AvaloniaUI/Avalonia/issues/12600).
